File: API_Run_of_Connectors/API_run_Connector_multiple_files.rb
# frozen_string_literal: true

# kenna run connectors
require 'rest-client'
require 'json'

@token = ARGV[0]
@folder = ARGV[1]
@connector_id = ARGV[2]
@file_extension = ARGV[3]

@LOCATOR_DELIMITER = ':'
@API_ENDPOINT_CONNECTOR = 'https://api.kennasecurity.com/connectors'
@headers = { 'content-type' => 'application/json', 'X-Risk-Token' => @token, 'accept' => 'application/json' }

@debug = false

def directory_exists?(directory)
  Dir.exist?(directory)
end

start_time = Time.now
output_filename = "kenna-runConnector_log-#{start_time.strftime('%Y%m%dT%H%M')}.txt"
log_output = File.open(output_filename, 'a+')
log_output << "Start time: time: #{Time.now}\n"

puts 'Directory not found' unless directory_exists?(@folder)

Dir.entries(@folder.to_s).each do |abspath|
  puts abspath
  next unless abspath.end_with? @file_extension.to_s

  fname = File.basename(abspath, @file_extension.to_s)

  conn_url = "#{@API_ENDPOINT_CONNECTOR}/#{@connector_id}/data_file?run=true"
  puts conn_url if @debug

  begin
    query_response = RestClient::Request.execute(
      method: :post,
      url: conn_url,
      headers: @headers,
      payload: {
        multipart: true,
        file: File.new("#{@folder}/" + abspath, 'rb')
      }
    )

    query_response_json = JSON.parse(query_response.body)

    puts query_response_json.fetch('success') if @debug

    running = true

    conn_check_url = "#{@API_ENDPOINT_CONNECTOR}/#{@connector_id}"

    while running

      sleep(15)
      conn_check = RestClient::Request.execute(
        method: :get,
        url: conn_check_url,
        headers: @headers
      )

      conn_check_json = JSON.parse(conn_check)['connector']
      puts conn_check_json if @debug
      running = conn_check_json.fetch('running')
    end
  rescue RestClient::UnprocessableEntity => e
    log_output = File.open(output_filename, 'a+')
    log_output << "UnprocessableEntity: #{e.message}... (time: #{Time.now}, start time: #{start_time})\n"
    log_output.close
    puts "UnprocessableEntity: #{e.message}"
  rescue RestClient::BadRequest => e
    log_output = File.open(output_filename, 'a+')
    log_output << "BadRequest: #{e.message}... (time: #{Time.now}, start time: #{start_time})\n"
    log_output.close
    puts "BadRequest: #{e.message}"
  rescue RestClient::Exception => e
    puts "i hit an exception #{e.message}"

    @retries ||= 0
    if @retries < @max_retries
      @retries += 1
      sleep(15)
      retry
    else
      log_output = File.open(output_filename, 'a+')
      log_output << "General RestClient error #{e.message}... (time: #{Time.now}, start time: #{start_time})\n"
      log_output.close
      puts "Exception: #{e.message}"
    end
  end
  log_output = File.open(output_filename, 'a+')
  log_output << "End time: time: #{Time.now}\n"
end

File: API_Run_of_Connectors/API_run_Connector_multiple_files.py
#!/usr/bin/env python

import sys, requests, json, os, time, urllib, pycurl

token = sys.argv[1]
folder = sys.argv[2]
connector_id = sys.argv[3]
file_extension = sys.argv[4]


LOCATOR_DELIMITER = ":"
API_ENDPOINT_CONNECTOR = "https://api.kennasecurity.com/connectors"
headers = {'content-type': 'application/json', 'X-Risk-Token': token}



for filename in os.listdir(folder):
    abspath = ''

    if filename.endswith(file_extension):
        pathname = os.path.join(folder, filename)
        abspath = os.path.abspath(pathname)
        try:

            conn_url = API_ENDPOINT_CONNECTOR + "/" + connector_id + "/data_file?run=true"

            print conn_url

            c = pycurl.Curl()
            c.setopt(c.URL, conn_url)
            c.setopt(c.POST, 1)
            c.setopt(c.HTTPPOST, [("file", (c.FORM_FILE, abspath))])
            c.setopt(pycurl.HTTPHEADER, ['content-type:application/json'])
            c.setopt(pycurl.HTTPHEADER, ['X-Risk-Token:'+token])
            c.setopt(c.VERBOSE, 0)
            c.perform()
            c.close()

            running = True

            while running:

                time.sleep(15)
                vuln_post = requests.get(API_ENDPOINT_CONNECTOR + "/" + connector_id, headers=headers)
                #print vuln_post.status_code
                response = vuln_post.json()
                print response
                running = response['connector']['running']


        except KeyError:
            print("unable to run connector")  
            raise   
          